Create per-buffer-theme.el

inigoserna opened the pull request:
**per-buffer-theme.el** is an Emacs library that automatically changes the global theme according to buffer name or major mode.

purcell commented:
Quick feedback:

- Public symbols shouldn't be abbreviated to `pbt`. Use `per-buffer-theme-...`.
- The `defvar`s should be `defcustom`s with a proper `defgroup`.
- Rather than hard-coding conventions for (and handling of) `:modes`, `:buffernames` etc., perhaps you should just allow the user to provide a function which can return a theme name, e.g. `per-buffer-theme-choose-theme-function`. This way, they can easily supply a function which returns different themes according to major mode, buffer name etc.

inigoserna commented:
First, thanks for your feedback, this is my first serious attemp to write an emacs library.

1. Is `per-buffer-theme/var-name` good enough? i.e. '/' character as separator between package name and variable name.
2. I'll take a look at how to do it
3. IMO this would be too complex, I created this package with the aim to provide an easy way to control theme depending on buffer name or major mode. 
The abstraction level you suggest is far from simple, at least with my elisp coding skills.
I think a map/hash table/dictionary with theme and buffer-name/major-mode is good enough for my uses.

Again, thanks for your feedback.

purcell commented:
> Is per-buffer-theme/var-name good enough? i.e. '/' character as separator between package name and variable name.

I think there's value in following the [elisp coding conventions](https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html). Clojure uses `/` to indicate namespaces, but elisp doesn't.

> IMO this would be too complex, I created this package with the aim to provide an easy way to control theme depending on buffer name or major mode. The abstraction level you suggest is far from simple, at least with my elisp coding skills. I think a map/hash table/dictionary with theme and buffer-name/major-mode is good enough for my uses.

It doesn't need to be complex. It would actually be simpler than the current code. If I get a few minutes I'll write some code to illustrate what I mean.

inigoserna commented:
I've pushed a couple of new commits:
- Now public functions and variables use `per-buffer-theme/` prefix
- Added defgroup and converted defvars to defcustoms

Note I decided to mantain '/' as separator between package name and symbol as elisp symbols looks clearer, specially if using `per-buffer-theme` instead of `pbt`.

> It doesn't need to be complex. It would actually be simpler than the current code. If I get a few minutes I'll write some code to illustrate what I mean.

I understand it, but IMO it is a very open solution.
What I'm not be able to see is how users could specify simple rules for changing themes depending on attributes like buffer name or mode. Of course advanced users could write such a function, but I think it's better to offer it.

Thanks again for your guidance.

purcell commented:
> Note I decided to mantain '/' as separator between package name and symbol as elisp symbols looks clearer, specially if using per-buffer-theme instead of pbt.

That's your prerogative, but IMO it's bad for the community whenever people deliberately ignore the coding standards. Yes, "/" might be more readable, but it's not what people expect.

> I understand it, but IMO it is a very open solution.
> What I'm not be able to see is how users could specify simple rules for changing themes depending on attributes like buffer name or mode. Of course advanced users could write such a function, but I think it's better to offer it.

I didn't explain it very clearly, but you can do both, and set the `alist`-based selection as the default "chooser" function:

```
(defcustom per-buffer-theme-choose-theme-function 'per-buffer-theme-choose-via-alist)

(defun per-buffer-theme-choose-via-alist ()
   ...)

...
     (let ((new-theme (funcall per-buffer-theme-choose-theme-function)))
        ....)
...
```

inigoserna commented:
> I didn't explain it very clearly, but you can do both, and set the alist-based selection as the default "chooser" function:

> ```
> (defcustom per-buffer-theme-choose-theme-function 'per-buffer-theme-choose-via-alist)
> 
> (defun per-buffer-theme-choose-via-alist ()
>    ...)
> 
> ...
>      (let ((new-theme (funcall per-buffer-theme-choose-theme-function)))
>         ....)
> ...```

I understand it, but my problem with this solution is that people would need to specify both function and data needed for that function, and it would make UI customization more difficult or imposible.

Of course we could specify a *defcustom* function such as `per-buffer-theme/choose-via-alist` to be used in the Emacs Custiomization System, but how could we add the option for data `per-buffer-theme/my-alist` or `per-buffer-theme/choose-via-major-mode` and `per-buffer-theme/my-major-modes`... The needed data for each customised function could be a different type, *alist* in first case, a simple list in the second, a *hash* in other...

Btw, I just pushed a new version with some minor modifications.

Thanks for your comments.

milkypostman commented:
All Steve is saying is that you would provide the function, and you would
provide a variable which is set to that function by default. Any would also
provide those variables, the data, for that function. So users have the
option to use the default and just provide the A-list, or if there really
savvy they can set the variable, which is by defaults it to your function,
To a function of their own choosing. This will not be any more complicated
because you already have to write the function to handle the A-list that
you are already using.

You can have a variable for both the a lists and the function, and if
someone chooses to replace the function the list don't matter. And it's
fine to do this it's just providing them a little more customization and if
there that advanced to be able to write a function for your code, their
danced enough to know that those list variables don't matter anymore.
On Mon, Apr 13, 2015 at 06:16 inigoserna <notifications@github.com> wrote:

> I didn't explain it very clearly, but you can do both, and set the
> alist-based selection as the default "chooser" function:
>
> (defcustom per-buffer-theme-choose-theme-function 'per-buffer-theme-choose-via-alist)
>
> (defun per-buffer-theme-choose-via-alist ()
>    ...)
>
> ...
>      (let ((new-theme (funcall per-buffer-theme-choose-theme-
>
> function)))
>         ....)
> ...```
>
>  I understand it, but my problem with this solution is that people would
> need to specify both function and data needed for that function, and it
> would make UI customization more difficult or imposible.
>
> Of course we could specify a *defcustom* function such as
> per-buffer-theme/choose-via-alist to be used in the Emacs Custiomization
> System, but how could we add the option for data per-buffer-theme/my-alist
> or per-buffer-theme/choose-via-major-mode and
> per-buffer-theme/my-major-modes... The needed data for each customised
> function could be a different type, *alist* in first case, a simple list
> in the second, a *hash* in other...
>
> Btw, I just pushed a new version with some minor modifications.
>
> Thanks for your comments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/milkypostman/melpa/pull/2672#issuecomment-92347535>.
>
